wagtail/tests/middleware.py

```diff
@@ -0,0 +1,8 @@
+from django.http import HttpResponseForbidden
+from django.utils.deprecation import MiddlewareMixin
+
+
+class BlockDodgyUserAgentMiddleware(MiddlewareMixin):
+    def process_request(self, request):
+        if not request.path.startswith('/admin/') and request.META.get('HTTP_USER_AGENT') == 'EvilHacker':
```

Comment:
I had to stare at this for a while to understand how this was working [in the new test](https://github.com/wagtail/wagtail/blob/92e5bf9ab5ecbb9ca7276b09bbfa8b9229bd91f1/wagtail/admin/tests/test_pages_views.py#L5266) since that test is making an admin preview request to a path that _does_ start with `/admin/`. Eventually I figured out that what's being tested is the "path" of the page being _previewed_ (in this case `/hello-world/`).

I'd be curious about real world use cases where Wagtail users might encounter non-200 responses from preview middleware. For example, say you had a middleware that redirects under certain conditions, say if a page has certain configuration settings. If you preview a page in that case, your "preview view" would end up redirecting. This makes logical sense but does feel a bit unexpected; as a Wagtail user I might expect some kind of interstitial page indicating that the preview had redirected.

Reply:
@chosak Yep, the `not request.path.startswith('/admin/')` needs to be there because the same HTTP headers are reused for the real request (to the admin view) and the faked request used to perform the preview, and we're testing the case where the latter triggers a middleware response. Can add a comment to clarify this if you think it'll help.

As far as I can see, the most common reason for middleware to return a non-200 response (or indeed to return early with its own 200 response rather than proceeding to the next middleware / view) is if the user has failed an authentication check - in this case it'll probably be clear to the user what has happened. Either way, returning that response as-is will definitely be an improvement over the current behaviour (i.e. forging ahead and attempting to serve the preview using the possibly-incomplete request object).

Reply:
That makes sense, thanks. A comment might help to indicate why that clause is necessary.

Reply:
That makes sense, thanks. A comment might help to indicate why that clause is necessary.